src/datastore.rs

```diff
@@ -0,0 +1,373 @@
+    pub fn len(&self) -> usize {
+        self.store.len()
+    }
+
+    /// Returns `true` if the store contains no elements.
+    pub fn is_empty(&self) -> bool {
+        self.store.is_empty()
+    }
+
+    /// An iterator visiting all paths in arbitrary order.
+    pub fn keys(&self) -> impl Iterator<Item = &PathBuf> {
+        self.store.keys()
+    }
+
+    /// An iterator visiting all content in arbitrary order.
+    pub fn values(&self) -> impl Iterator<Item = &Vec<u8>> {
+        self.store.values()
+    }
+
+    /// An iterator visiting all content mutably in arbitrary order.
+    pub fn values_mut(&mut self) -> impl Iterator<Item = &mut Vec<u8>> {
+        self.store.values_mut()
+    }
+
+    /// Tries to insert a path-content pair into the store.
+    ///
+    /// If the store did not have this path present, `None` is returned.
+    ///
+    /// If the store did have this path present, the content is updated, and the old
+    /// content is returned. The path is not updated, though.
+    ///
+    /// Returns a [`DataStoreError`] if:
+    /// 1. The path is empty.
+    /// 2. The path is absolute.
+    /// 3. Any of the path's ancestors is already tracked in the store, implying
+    ///    the path to be nested under a file.
+    pub fn try_insert(
+        &mut self,
+        k: PathBuf,
+        v: Vec<u8>,
+    ) -> Result<Option<Vec<u8>>, DataStoreError> {
+        if k.as_os_str().is_empty() {
+            return Err(DataStoreError::EmptyPath);
+        }
+        if k.is_absolute() {
+            return Err(DataStoreError::PathIsAbsolute);
+        }
+        for ancestor in k.ancestors() {
+            if !ancestor.as_os_str().is_empty() && self.store.contains_key(ancestor) {
+                return Err(DataStoreError::DirUnderFile);
+            }
+        }
+        Ok(self.store.insert(k, v))
+    }
+
+    /// Removes a path from the store, returning the content at the path if the path
+    /// was previously in the store.
+    pub fn remove(&mut self, k: &Path) -> Option<Vec<u8>> {
+        self.store.remove(k)
+    }
+
+    /// Returns a reference to the content corresponding to the path.
+    pub fn get(&self, k: &Path) -> Option<&Vec<u8>> {
+        self.store.get(k)
+    }
+
+    /// Returns a mutable reference to the content corresponding to the path.
+    pub fn get_mut(&mut self, k: &Path) -> Option<&mut Vec<u8>> {
+        self.store.get_mut(k)
+    }
+
+    /// An iterator visiting all path-content pairs in arbitrary order.
+    pub fn iter(&self) -> impl Iterator<Item = (&PathBuf, &Vec<u8>)> {
+        self.store.iter()
+    }
+
+    /// An iterator visiting all path-content pairs in arbitrary order, with mutable
+    /// references to the content.
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = (&PathBuf, &mut Vec<u8>)> {
+        self.store.iter_mut()
+    }
+}
+
+/// There are no methods for getting mutable references to images to prevent circumvention of the PNG validity check.
+impl ImageStore {
```

Comment:
I think we should be able to combine these into a single type.

From what I can tell, there are two places where these two types differ: in how they do the initial data loading, and in how they handle insertion.

So what we can imagine is some shared `DataStore` type that allows customization of this behaviour. One option would be to use a trait, like:

```rust
trait DataType {
    type Error;
    fn try_load(&self, path: &Path) -> Result<HashMap<PathBuf, Vec<u8>>>;
    fn try_insert(
        &self,
        collection: &mut HashMap<PathBuf, Vec<u8>>,
        path: PathBuf,
        data: Vec<u8>,
    )  -> Result<Option<Vec<u8>>, ImageStoreError>;
}


struct Data;
impl DataType for Data { 
    // impl here
}

struct Image;

impl DataType for Image {
    // image specific code here
}

struct Store<T> {
    store: HashMap<PathBuf, Vec<u8>>,
    kind: T,
}

type DataStore = Store<Data>;
type ImageStore = Store<Image>;
```

but this might be overkill? Another simpler option would be to have separate `try_new_images` and `try_new_data` methods, and then just store a `bool` indicating whether we've loaded images or data, and switch on this bool in `try_insert` in order to do validation correctly?

In any case these are just a couple possible options, I'm sure you can think of others. :)